### DISCUSSION TITLE
How WinUI3 refreshes the UI thread in other threads

### DISCUSSION THREAD

**Discussion by far1shx** [microsoft/microsoft-ui-xaml#8732] [Q&A]
How WinUI3 refreshes the UI thread in other threads
I've been stuck with this question for a long time, hope someone answers thank you

**Comment by AndrewKeepCoding** [comment]
The recommended way is to use the `DispatcherQueue`. Let me show you some examples:

```cs
public sealed partial class MainPage : Page
{
    private DispatcherQueue _dispatcherQueue;

    public MainPage()
    {
        this.InitializeComponent();
        this.Loaded += MainPage_Loaded;
        _dispatcherQueue = DispatcherQueue.GetForCurrentThread();
    }

    private async void MainPage_Loaded(object sender, Microsoft.UI.Xaml.RoutedEventArgs e)
    {
        // This works because this event handler is called on the UI thread.
        Button button = new();

        // This throws an exception because the button is not created on the UI thread.
        await Task.Run(() =>
        {
            Button button = new();
        });

        // This works because the DispatcherQueue is used to create the button on the UI thread.
        this.DispatcherQueue.TryEnqueue(() =>
        {
            Button button = new();
        });

        // This works because the DispatcherQueue is used to create the button on the UI thread.
        _dispatcherQueue.TryEnqueue(() =>
        {
            Button button = new();
        });
    }
}
```

**Reply by far1shx** [reply]
Okay thank you

**Comment by DarranRowe** [comment]
The big thing to remember is that the desktop windowing system means that you need a loop on UI threads to handle messages. Usually it is a loop along the lines of:
```
MSG msg{};
while(GetMessage(&msg, nullptr, 0, 0))
{
    //Potentially other functionality here.
    DispatchMessage(&msg);
}
```
This means that there is one way to get back to the UI thread from any other thread if you have a window handle. There is also the concept of a message only window, which is a window that can never be made visible and only exists to receive messages.
Now, another thing to remember is that COM, which WinRT is, has the concept of apartments. The UI thread is normally set as STA (single threaded apartment). While this has a lot of implications, the biggest thing of importance is that COM uses the message handling loop and a message only window to implement important functionality. The reason why this is important is that the COM runtime knows how to find this, and if you do it right, you can tell COM to run code in a specific apartment. The function CoGetObjectContext can be used to obtain a pointer to the IContextCallback interface. This interface can then be used to run a function in the apartment that the context belongs to. This should mean that it sends a message to the COM runtime's message only window with a pointer to the function that it needs to execute, and the handler for this message just calls the function that the message points to.
While this is a talk about native code, these services would be available to .NET too.

**Reply by far1shx** [reply]
Okay thank you
